extensions/amp-ad/0.1/amp-ad-api-handler.js

```diff
@@ -121,21 +130,32 @@ export class AmpAdApiHandler {
       // creative as it will not expect having to send the render-start message.
       this.iframe_.style.visibility = 'hidden';
     }
-    listenForOnce(this.iframe_, 'render-start', () => {
-      if (!this.iframe_) {
-        return;
-      }
-      if (this.baseInstance_.renderStartResolve_) {
-        this.baseInstance_.renderStartResolve_();
-        this.baseInstance_.renderStartResolve_ = null;
-      }
-      this.iframe_.style.visibility = '';
-    }, this.is3p_);
+
+    this.renderStartPromise_ =
+        listenForOncePromise(this.iframe_, 'render-start', this.is3p_);
+
     this.viewer_.onVisibilityChanged(() => {
       this.sendEmbedInfo_(this.baseInstance_.isInViewport());
     });
     this.element_.appendChild(this.iframe_);
-    return loadPromise(this.iframe_);
+    return loadPromise(this.iframe_).then(() => {
+      this.renderStartTimeoutPromise_ = timerFor(this.baseInstance_.win)
+          .timeoutPromise(200, this.renderStartPromise_,
+          'render-start-event timed out');
+      this.renderStartTimeoutPromise_.then(() => {
+        //TODO: report performance
+        if (this.iframe_) {
+          this.iframe_.style.visibility = '';
+        }
+      }).catch(error => {
+        //TODO: report performance
+        if (this.iframe_) {
+          this.iframe_.style.visibility = '';
+        }
+        dev().error(error);
```

Comment:
You may also want to rewrite it this way:

```
this.renderStartTimeoutPromise_.catch(error = > {
  dev().error(TAG, error);
}).then(() => {
  if (this.iframe_) {
    this.iframe_.style.visibility = '';
  }
});
```

Further, it's most likely incorrect to sub-chain the promise here and then return `return this. renderStartTimeoutPromise_;`  - basically what happens is that you will return loading promise before you completed all loading actions.

And finally, do you actually need an instance var? Why not just do:
```
return timer
    .timeoutPromise(200, this.renderStartPromise_, ...)
    .catch(error => { /* report error */ })
    .then(() => { /* complete loading/init re:visibility */});
```